### PR TITLE
Add server connector connection to resource signature 

### LINKED ISSUE #4183: [HTTP] Refactor resource signature
At the moment, we use Request and Response structs to represent both inbound and outbound request/response. However, there is a slight problem with that,

1) some of the functions associated to these structs no valid for both inbound and outbound scenarios. 
2) Response is the struct that send the back response to the client but ideally this should have done using the server connector (we don't have it at the moment but want to show it in the visual diagram).

### LINKED ISSUE #4245: [HTTP] Restructure resource signature
We need to do the subject as described in the doc [1]. Please refer to V3's approach 3.

[1] https://docs.google.com/document/d/12iInthCIFyJclGQDVhG52T8H7eMbyyyHRbsQil0KJ_I/edit#heading=h.uyy2u1vvlex3

### PR DESCRIPTION
**Pull request by chamil321**
PR introduces 
- server connector connection as a param in resource signature. 
- respond native function to send back the response to the caller. 
- test cases and migrate existing test cases
- fixes #4183 
- fixes #4245 
```ballerina
service<http> helloWorld {
    resource sayHello (http:Connection con, http:Request req) {
        http:Response res = {};
        res.setStringPayload("Hello, World!");
        _ = con.respond(res);
    }
}
```
PR for tools distribution - https://github.com/ballerinalang/tools-distribution/pull/231

### CODE REVIEW

File: modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java [hunk 1]
```diff
@@ -755,8 +752,7 @@ public static void setHeadersToTransportMessage(HTTPCarbonMessage cMsg, BStruct
             return null;
         }
         HttpHeaders removedHeaders = new DefaultHttpHeaders();
-        if (struct.getNativeData(Constants.OUTBOUND_RESPONSE) == null
-                && !struct.getStringField(Constants.RESPONSE_SERVER_INDEX).equals("")) {
+        if (!struct.getStringField(Constants.RESPONSE_SERVER_INDEX).equals("")) {
```
Comment (anupama-pathirage) [inline modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java]: Can getStringField return null here? Also isEmpty() would be better

Comment (chamil321) [reply modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java]: getStringField is initialized as an empty string so it won't be a null. I'll change equals() to isEmpty()

File: modules/ballerina-test/src/test/resources/test-src/statements/services/nativeimpl/request/request-native-function.bal [hunk 2]
```diff
@@ -168,14 +173,15 @@ service<http> helloServer {
         int size =  lengthof valueArr;
         string header = valueArr[2].value;
 
+        http:Response res = {};
         res.setJsonPayload({headerValue:header, size:size});
-        _ = res.send();
+        _ = con.respond(res);
```
Comment (shafreenAnfar) [inline modules/ballerina-test/src/test/resources/test-src/statements/services/nativeimpl/request/request-native-function.bal]: Shall we rename con to conn ?

Comment (chamil321) [reply modules/ballerina-test/src/test/resources/test-src/statements/services/nativeimpl/request/request-native-function.bal]: ack

Comment (chamil321) [reply modules/ballerina-test/src/test/resources/test-src/statements/services/nativeimpl/request/request-native-function.bal]: Ack

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package ballerina.net.http;
+
+@Description { value:"Sends a 100-continue response to the client."}
+@Param { value:"conn: The server connector connection" }
```
Comment (pubudu91) [inline modules/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: I feel that this description may not be user friendly. Will it be possible to improve this? Is the user likely to understand what we mean when we just simply say "The server connector connection"?

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/natives.bal [hunk 4]
```diff
@@ -1,16 +1,36 @@
 package ballerina.net.http;
 
-@Description { value:"Represents an HTTP request message"}
-@Field {value:"host: The server host name"}
+@Description { value:"Represents the HTTP server connection"}
```
Comment (pubudu91) [inline modules/ballerina-http/src/main/ballerina/ballerina/net/http/natives.bal]: Again, is it correct to refer to this Connection struct as an "HTTP server connection"?

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal (hunk 3 above)
Comment (pubudu91) [reply modules/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: Also, isn't there any other place we can put this function?

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/natives.bal [hunk 5]
```diff
@@ -1,16 +1,36 @@
...
+@Field {value:"remoteHost: The server host name"}
 @Field {value:"port: The server port"}
+public struct Connection {
+	string remoteHost;
+	int port;
+}
+
+@Description { value:"Sends response to the caller"}
+@Param { value:"conn: The server connector connection" }
```
Comment (pubudu91) [inline modules/ballerina-http/src/main/ballerina/ballerina/net/http/natives.bal]: Same concern as the previous comment on the terminology

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal (hunk 3 above)
Comment (chamil321) [reply modules/ballerina-http/src/main/ballerina/ballerina/net/http/connection.bal]: keeping the same description as user can get the rough idea about the conn with it.

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/natives.bal (hunk 5 above)
Comment (chamil321) [reply modules/ballerina-http/src/main/ballerina/ballerina/net/http/natives.bal]: keeping the same description as user can get the rough idea about the conn with it.

File: modules/ballerina-http/src/main/ballerina/ballerina/net/http/natives.bal (hunk 4 above)
Comment (chamil321) [reply modules/ballerina-http/src/main/ballerina/ballerina/net/http/natives.bal]: Changed it as HTTP server connector connection

**Pull request merged**